source/common/network/io_socket_error_impl.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "envoy/api/io_error.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Network {
+
+class IoSocketError : public Api::IoError {
+public:
+  explicit IoSocketError(int sys_errno) : errno_(sys_errno) {}
+
+  ~IoSocketError() override {}
+
+  Api::IoError::IoErrorCode getErrorCode() const override;
+  std::string getErrorDetails() const override;
+
+private:
+  int errno_;
+};
+
+// IoErrorCode::Again is used frequently. Define it to be a singleton to avoid frequent memory
+// allocation of such instance. If this is used, IoHandleCallResult has to be instantiated with
+// deleter deleteIoError() below to avoid deallocating memory for this error.
+class IoSocketEagain : public IoSocketError {
+public:
+  IoSocketEagain() : IoSocketError(EAGAIN){};
+};
+
+inline IoSocketEagain* getIoSocketEagainInstance() {
+  static auto* kInstance = new IoSocketEagain();
+  return kInstance;
+}
+
+// Deallocate memory only if the error is not IoSocketErrorAgain.
+inline void deleteIoError(Api::IoError* err) {
+  ASSERT(err != nullptr);
+  if (err != getIoSocketEagainInstance()) {
```

Comment:
rather than calling getIoSocketEgainInstance() you could just compare the code to AGAIN (if you really don't like free() being a virtual method on IoError per the above comment.